zkvm/src/predicate.rs

```diff
@@ -27,21 +29,38 @@ pub enum Predicate {
     /// constructing aggregated signature.
     Key(VerificationKey),
 
-    /// Representation of a predicate as commitment to a program and blinding factor.
-    Program(Program, Vec<u8>),
-
-    /// Disjunction of n predicates.
-    Or(PredicateDisjunction),
+    /// Taproot Merkle tree.
+    Tree(PredicateTree),
 }
 
 #[derive(Clone, Debug)]
-pub struct PredicateDisjunction {
-    preds: Vec<Predicate>,
-    // We precompute the disjunction predicate when composing it via `Predicate::or()`,
+pub struct PredicateTree {
+    pub progs: Vec<Program>,
+    pub key: VerificationKey,
+    blinding: [u8; 32],
+    // We precompute the tweaked aggregated signing key when composing it via `Predicate::tree()`,
     // so that we can keep `to_point`/`encode` methods non-failable across all types.
     precomputed_point: CompressedRistretto,
 }
 
+#[derive(Clone, Debug, Default)]
+pub struct CallProof {
+    // Signing key
+    pub signing_key: VerificationKey,
+
+    // Bit pattern indicating neighbor position
+    pub positions: u32,
```

Comment:
`MerkleNeighbor` already encodes the position via Left/Right variants. This 32-bit pattern can be kept at the serialization level only. In this struct you can keep the neighbors and verification key only.

Later, to avoid reallocations, you can change the proof-verifying code to take iterator of `MerkleNeighbor`s, so we don't have to allocate a Vec of them, but just parse them on the fly from the serialized callproof.